utils/pretty_printers.cc

```diff
@@ -11,16 +11,21 @@
 namespace utils {
 
 std::ostream& operator<<(std::ostream& os, pretty_printed_data_size data) {
-    static constexpr const char *suffixes[] = {" bytes", "kB", "MB", "GB", "TB", "PB"};
-
-    unsigned exp = 0;
-    while ((data._size >= 1000) && (exp < sizeof(suffixes))) {
-        exp++;
-        data._size /= 1000;
+    static constexpr auto suffixes = {" bytes", "kB", "MB", "GB", "TB", "PB"};
+
+    std::string_view suffix;
+    uint64_t size = data._size;
+    uint64_t next_size = size;
+    for (auto s : suffixes) {
+        suffix = s;
+        size = next_size;
+        next_size = size / 1000;
+        if (next_size == 0) {
+            break;
+        }
     }
 
-    os << data._size << suffixes[exp];
-    return os;
+    return os << size << suffix;
```

Comment:
I think your new code is correct, but you didn't have to completely rewrite this function... I think the original code simply had a trivial bug - it should have "exp < sizeof(suffixes) - 1" as the end of the loop. That's it!

Reply:
right. I chose to reimplement it instead patching it like you suggest. The reason it was buggy was that the off by one bug was difficult to identify. So I think a better way to fix it is to implement it a different way as I explained in the commit message.

Reply:
For me, the new code was harder to understand. I guess it's a matter of style.

The new code is also pretty inefficient (silly loop over character pointers converting all of them to std::string_view), but much of it can be easily solved (although your version still has a more assignments and copies of the "size" variable). Anyway, as I said, I approve. I think the unit test is more important than the specific implementation - any implementation can be wrong in a hard-to-spot way if not tested.

Reply:
i will change the assignment part. you mentioned this already in another comment. i did read it.